src/config/proxy.rs

```diff
@@ -0,0 +1,169 @@
+use http::uri::{InvalidUri, Uri};
+use hyper_proxy::{Custom, Intercept, Proxy, ProxyConnector};
+use std::collections::HashSet;
+
+fn from_env(key: &str) -> Option<String> {
+    std::env::var(key.to_string())
+        .ok()
+        .or_else(|| std::env::var(key.to_lowercase()).ok())
+}
+
+struct NoProxyCache(HashSet<Uri>);
+
+impl From<&HashSet<String>> for NoProxyCache {
+    fn from(hashset: &HashSet<String>) -> Self {
+        Self(
+            hashset
+                .iter()
+                .filter_map(|uri| uri.parse::<Uri>().ok())
+                .collect(),
+        )
+    }
+}
+
+impl NoProxyCache {
+    fn matches(&self, scheme: Option<&str>, host: Option<&str>, port: Option<u16>) -> bool {
+        for uri in self.0.iter() {
+            if uri.scheme_str() == scheme && uri.host() == host && uri.port_u16() == port {
+                return true;
+            }
+        }
```

Comment:
I've observed that different clients implement `no_proxy` differently. I think what you have here is good, but what would you think of expanding it to cover a few more cases as suggested by https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/ ?

> Use * to match all hosts.
> Strip leading dots (.) and match against domain suffixes.
> Support CIDR block matching.

I'm noting that they also recommend checking for lowercased versions before uppercased. I don't have a strong opinion on that.

Reply:
Considering the "complexity" of this `NO_PROXY` proposed by the people from gitlab (good article btw) and the fact that it's propably going to be reimplemented by someone else, would it be worth making a separate crate for it? a `no-proxy-parser` of some sort? Or even a `proxy-config-parser` 🤔

Reply:
Just noting that we talked outside and a separate crate for handling no_proxy matching seems reasonable given the complexity.